src/components/Profile/useUserInfoForm.ts

```diff
@@ -28,63 +60,95 @@ export default function useUserInfoForm({
 }: Props) {
   const [username, setUsername] = useState(existingUsername ?? '');
   const [usernameError, setUsernameError] = useState('');
+  const usernameFieldIsDirty = useRef(false);
 
   const [bio, setBio] = useState(existingBio ?? '');
 
   // Generic error that doesn't belong to username / bio
   const [generalError, setGeneralError] = useState('');
   const updateUser = useUpdateUser();
+  const createUser = useCreateUser();
+  const authPayloadQuery = useAuthPayloadQuery();
 
   const handleCreateOrEditUser = useCallback(async () => {
     setGeneralError('');
 
-    // -------------- client-side checks --------------
-    const usernameError = validate(username, [
-      required,
-      minLength(2),
-      maxLength(20),
-      alphanumericUnderscores,
-      noConsecutivePeriodsOrUnderscores,
-    ]);
-
     if (usernameError) {
-      setUsernameError(usernameError);
-      return;
+      return { success: false };
     }
 
     if (bio.length > BIO_MAX_CHAR_COUNT) {
       // No need to handle error here, since the form will mark the text as red
-      return;
+      return { success: false };
     }
-    // ------------ end client-side checks ------------
 
     try {
-      await updateUser(userId, username, bio);
-
+      if (userId) {
+        await updateUser(userId, username, bio);
+      } else {
+        if (!authPayloadQuery) {
+          throw new Error('Auth signature for creating user not found');
+        }
+        await createUser(authPayloadQuery, username, bio);
+      }
       onSuccess(username);
+      return { success: true };
     } catch (error: unknown) {
       if (error instanceof Error) {
         setGeneralError(formatError(error));
       }
+      return { success: false };
     }
-  }, [username, bio, updateUser, userId, onSuccess]);
+  }, [usernameError, bio, userId, authPayloadQuery, onSuccess, username, updateUser, createUser]);
 
-  const handleClearUsernameError = useCallback(() => {
-    setUsernameError('');
+  const handleUsernameChange = useCallback((username: string) => {
+    setUsername(username);
+    usernameFieldIsDirty.current = true;
   }, []);
 
+  const isUsernameAvailableFetcher = useIsUsernameAvailableFetcher();
+
+  const debouncedUsername = useDebounce(username, 500);
```

Comment:
wait a bit before asking the server whether a username is taken (to prevent spamming it)